diff --git a/murata/files/wifi_mod_para_murata.conf b/murata/files/wifi_mod_para_murata.conf
index ac98f82..3f39cba 100644
--- a/murata/files/wifi_mod_para_murata.conf
+++ b/murata/files/wifi_mod_para_murata.conf
@@ -153,14 +153,15 @@ SD9098_1 = {
 
 SD8978 = {
 	cfg80211_wext=0xf
-	wfd_name=p2p
 	max_vir_bss=1
+	sta_name=wlan
 	cal_data_cfg=none
 	drv_mode=7
 	ps_mode=2
 	auto_ds=2
 	fw_name=nxp/sdiouartiw416_combo_v0.bin
-	txpwrlimit_cfg=nxp/txpower_US.bin
+	txpwrlimit_cfg=nxp/txpower_EU.bin
+	init_hostcmd_cfg=nxp/ed_mac.bin
 }
 
 SD8801 = {
